vortex-array/src/builders/primitive.rs

```diff
@@ -48,23 +49,50 @@ impl<T: NativePType> PrimitiveBuilder<T> {
     }
 
     pub fn patch(&mut self, patches: Patches, starting_at: usize) -> VortexResult<()> {
-        let (array_len, offset, indices, values) = patches.into_parts();
+        let (array_len, indices_offset, indices, values) = patches.into_parts();
         assert!(starting_at + array_len == self.len());
         let indices = indices.into_primitive()?;
         let values = values.into_primitive()?;
         let validity = values.validity_mask()?;
         let values = values.as_slice::<T>();
         match_each_unsigned_integer_ptype!(indices.ptype(), |$P| {
-            for (compressed_index, decompressed_index) in indices.into_primitive()?.as_slice::<$P>().into_iter().enumerate() {
-                let decompressed_index = *decompressed_index as usize;
-                let out_index = starting_at + decompressed_index - offset;
-                if validity.value(compressed_index) {
-                    self.values[out_index] = values[compressed_index]
-                } else {
-                    self.nulls.set_bit(out_index, false)
-                }
+            if !matches!(validity, Mask::AllTrue(_)) {
```

Comment:
I think you can only do this will nullability not validity?

Reply:
Mmm. I was thinking that the validity of a position which will be patched is true, but I think you're right that the unpatched values could indeed be invalid.